x-pack/plugins/audit_trail/server/client/audit_trail_client.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { Subject } from 'rxjs';
+import { KibanaRequest, Auditor, AuditableEvent } from 'src/core/server';
+import { AuditEvent } from '../types';
+
+import { SecurityPluginSetup } from '../../../security/server';
+import { SpacesPluginSetup } from '../../../spaces/server';
+
+interface Deps {
+  getCurrentUser: SecurityPluginSetup['authc']['getCurrentUser'];
+  getSpaceId: SpacesPluginSetup['spacesService']['getSpaceId'];
+}
+
+export class AuditTrailClient implements Auditor {
+  private scope?: string;
+  constructor(
+    private readonly request: KibanaRequest,
+    private readonly event$: Subject<AuditEvent>,
+    private readonly deps: Deps
+  ) {}
+
+  public withScope(name: string) {
+    if (this.scope !== undefined) {
+      throw new Error(`AuditTrail scope is already set to: ${this.scope}`);
+    }
+    this.scope = name;
+  }
```

Comment:
As there is only one auditor for a specific request, that means we can only have a single scope for a request's whole lifecycle. Is that the expected granularity, or just a limitation caused by the fact that changing the scope of the auditor would just have side effect with asynchronous calls?

to adapt an example from the FTR test plugin, this is something that can will just not be able to do:

```ts
 router.get(
      { path: '/audit_trail_test/context/as_current_user', validate: false },
      async (context, request, response) => {
        context.core.auditor.withScope('audit_trail_test/context/as_current_user/ping');
        await context.core.elasticsearch.legacy.client.callAsCurrentUser('ping');
        context.core.auditor.withScope('audit_trail_test/context/as_current_user/pong'); // BOOM.
        await context.core.elasticsearch.legacy.client.callAsCurrentUser('pong');
        return response.noContent();
      }
    );
```

Reply:
> As there is only one auditor for a specific request, that means we can only have a single scope for a request's whole lifecycle. Is that the expected granularity, or just a limitation caused by the fact that changing the scope of the auditor would just have side effect with asynchronous calls?

It's the desired behavior. @jportner pointed out that we are interested in the root cause of an event, so I rolled back the possibility to stack 'scopes'.